### PR TITLE
Add New Feature: support CRI Interface docking docker 

### PR DESCRIPTION
**Pull request by arcanique**
Add Feature CRI support docker:

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind feature

**What this PR does / why we need it**:
>
> support CRI Interface docking docker 

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: edge/pkg/edged/edged.go [hunk 1]
```diff
@@ -26,8 +26,12 @@ package edged
 import (
 	"encoding/json"
 	"fmt"
+	"k8s.io/kubernetes/pkg/kubelet/dockershim"
+	"k8s.io/kubernetes/pkg/kubelet/server/streaming"
```
[the remark this answers is not in the document]
Comment (arcanique) [reply edge/pkg/edged/edged.go]: ok

File: edge/pkg/edged/edged.go [hunk 2]
```diff
@@ -473,6 +540,42 @@ func newEdged() (*edged, error) {
 				conf.remoteImageEndpoint = conf.remoteRuntimeEndpoint
 			}
 		}
+		//create and start the docker shim running as a grpc server
+		if conf.remoteRuntimeEndpoint == DockerShimEndpoint || conf.remoteRuntimeEndpoint == DockerShimEndpointDeprecated {
```
Comment (qizha) [inline edge/pkg/edged/edged.go]: why we need to support DockerShimEndpointDeprecated ?

File: edge/pkg/edged/edged.go [hunk 3]
```diff
@@ -473,6 +540,42 @@ func newEdged() (*edged, error) {
...
+			}
+
+			server := dockerremote.NewDockerServer(conf.remoteRuntimeEndpoint, ds)
+			if err := server.Start(); err != nil {
+				return nil, err
+			}
+
+		}
+		ed.clusterDNS = convertStrToIP(conf.clusterDNS)
+		ed.dnsConfigurer = kubedns.NewConfigurer(recorder, nodeRef, ed.nodeIP, ed.clusterDNS, conf.clusterDomain, ResolvConfDefault) //TODO
```
Comment (qizha) [inline edge/pkg/edged/edged.go]: what will happen if there is no kubedns available?

File: edge/pkg/edged/edged.go (hunk 2 above)
Comment (arcanique) [reply edge/pkg/edged/edged.go]: Dockershim is still supported, and there will be a waring log

Comment (arcanique) [reply edge/pkg/edged/edged.go]: yes，

> there will be no DNS available in most of the edge cases, can we handle that with this code? I saw the CNI initialization in the code, does that impact our current non-CNI situation?

We can configure cluster dns to be None and let ‘’clusterDNS‘’ take effect. And I think the cni initlization of the code is in order to fill the parameters and greatly make sure that there is no null pointer error for some method call

File: edge/pkg/edged/edged.go (hunk 3 above)
Comment (arcanique) [reply edge/pkg/edged/edged.go]: How to understand kubedns，do you mean the clusterDNS ?  The "clusterDNS" works with the dnsPolicy configuration at the deployment file

File: edge/conf/edge.yaml [hunk 4]
```diff
@@ -50,6 +50,10 @@ edged:
     remote-image-endpoint: /var/run/containerd/containerd.sock
     runtime-request-timeout: 2
     podsandbox-image: k8s.gcr.io/pause
+    cgroup-driver: systemd
+    node-ip: ""
```
Comment (sids-b) [inline edge/conf/edge.yaml]: Why is node-ip made a flag ?

File: edge/pkg/edged/edged.go [hunk 5]
```diff
@@ -26,11 +26,15 @@ package edged
...
 	"encoding/json"
 	"fmt"
+	"net"
 	"net/http"
 	"os"
+	"strings"
 	"sync"
 	"time"
 
+	"github.com/golang/glog"
```
[the remark this answers is not in the document]
Comment (arcanique) [reply edge/pkg/edged/edged.go]: ok,I will modify this

File: edge/pkg/edged/edged.go [hunk 6]
```diff
@@ -185,6 +219,12 @@ type edged struct {
 	configMapStore cache.Store
 	workQueue      queue.WorkQueue
 	clcm           clcm.ContainerLifecycleManager
+	//edged cgroup drive
```
[the remark this answers is not in the document]
Comment (arcanique) [reply edge/pkg/edged/edged.go]: ok,I will fix this problem

File: edge/conf/edge.yaml (hunk 4 above)
Comment (arcanique) [reply edge/conf/edge.yaml]: When multiple network planes exist on the host, the user is allowed to specify the IP address of a plane as the node ip. If not specified, the loopback address is used.

Comment (sids-b) [reply edge/conf/edge.yaml]: I think this flag is used to specify which IP to be considered ?
https://github.com/kubeedge/kubeedge/blob/master/edge/conf/edge.yaml#L38

[the remark this answers is not in the document]
Comment (arcanique) [reply edge/conf/edge.yaml]: yes，I have set the default values for these parameters in edged.go, just on line 372. And if the user didn't specified these flags, the program still works

Comment (arcanique) [reply edge/conf/edge.yaml]: yes， but I think sometimes  there will be several IPs on the same interface maybe. I think use net interface is not a better way to get node-ip. So, when I made this piece, I used node-ip.

Comment (sids-b) [reply edge/conf/edge.yaml]: Not quite sure about this. cc @qizha @kevin-wangzefeng @rohitsardesai83

Comment (qizha) [reply edge/conf/edge.yaml]: so kubelet have both, let's keep the same way with kubelet.

File: edge/conf/edge.yaml [hunk 7]
```diff
@@ -50,6 +50,10 @@ edged:
...
+    cluster-dns: ""
+    cluster-dmain: ""
```
Comment (rohitsardesai83) [inline edge/conf/edge.yaml]: dmain -> domain

File: edge/pkg/edged/edged.go [hunk 8]
```diff
@@ -120,11 +127,37 @@ const (
...
+	//DockerEndpoint gives the default endpoint for docker engine
+	DockerEndpoint = "unix:///var/run/docker.sock"
+	//DockerShimEndpoint gives the default endpoint for Docker shim runtime
+	DockerShimEndpoint = "unix:///var/run/dockershim.sock"
+	//DockerShimEndpointDeprecated this is the deprecated dockershim endpoint
+	DockerShimEndpointDeprecated = "/var/run/dockershim.sock"
+	//DockershimRootDir givesthe default path to the dockershim root directory
+	DockershimRootDir = "/var/lib/dockershim"
+	//HairpinMod only use forkubenetNetworkPlugin.Currently not working
+	HairpinMod = kubeletinternalconfig.HairpinVeth
```
Comment (rohitsardesai83) [inline edge/pkg/edged/edged.go]: HairpinMod -> HairpinMode ?

File: edge/pkg/edged/edged.go [hunk 9]
```diff
@@ -120,11 +127,37 @@ const (
...
+	//NonMasqueradeCIDR only use forkubenetNetworkPlugin.Currently not working
+	NonMasqueradeCIDR = "10.0.0.1/8"
+	//croupName used for check if the cgroup is mounted.(default "")
+	croupName = ""
```
Comment (rohitsardesai83) [inline edge/pkg/edged/edged.go]: croupName -> cgroupName ?

File: edge/pkg/edged/edged.go [hunk 10]
```diff
@@ -206,6 +247,10 @@ type Config struct {
 	remoteImageEndpoint      string
 	RuntimeRequestTimeout    metav1.Duration
 	PodSandboxImage          string
+	cgoroupDriver            string
```
Comment (rohitsardesai83) [inline edge/pkg/edged/edged.go]: cgoroupDriver -> cgroupDriver

File: edge/conf/edge.yaml (hunk 4 above)
Comment (sids-b) [reply edge/conf/edge.yaml]: OK

File: edge/conf/edge.yaml (hunk 7 above)
Comment (arcanique) [reply edge/conf/edge.yaml]: ok，I will modify this

File: edge/pkg/edged/edged.go (hunk 8 above)
Comment (arcanique) [reply edge/pkg/edged/edged.go]: ok，I will modify this

File: edge/pkg/edged/edged.go (hunk 9 above)
Comment (arcanique) [reply edge/pkg/edged/edged.go]: ok，I will modify this

File: edge/pkg/edged/edged.go [hunk 11]
```diff
@@ -90,6 +74,29 @@ import (
...
+	"github.com/kubeedge/beehive/pkg/common/util"
+	"github.com/kubeedge/beehive/pkg/core"
+	"github.com/kubeedge/beehive/pkg/core/context"
+	"github.com/kubeedge/beehive/pkg/core/model"
+	"github.com/kubeedge/kubeedge/edge/pkg/common/modules"
+	"github.com/kubeedge/kubeedge/edge/pkg/edged/apis"
+	"github.com/kubeedge/kubeedge/edge/pkg/edged/clcm"
+	"github.com/kubeedge/kubeedge/edge/pkg/edged/containers"
+	"github.com/kubeedge/kubeedge/edge/pkg/edged/dockertools"
+	edgeImages "github.com/kubeedge/kubeedge/edge/pkg/edged/images"
```
Comment (rohitsardesai83) [inline edge/pkg/edged/edged.go]: edgeImages -> edgeimages

File: edge/pkg/edged/edged.go (hunk 10 above)
Comment (arcanique) [reply edge/pkg/edged/edged.go]: ok，I will modify this

File: edge/pkg/edged/edged.go [hunk 12]
```diff
@@ -120,11 +127,37 @@ const (
...
+	croupName = ""
+	// PluginName gives the plugin name.(default "",use noop plugin)
+	pluginName = ""
+	//PluginBinDir gives the dir of cni plugin executable file
+	pluginBinDir = "/opt/cni/bin"
+	// PluginConfDir gives the dir of cni plugin confguration file
+	pluginConfDir = "/etc/cni/net.d"
+	//MTU give the default maximum transmission unit of  net interface
+	mtu = 1500
+	// redirectContainerStream decide whether to redirect the dontainer stream
```
Comment (rohitsardesai83) [inline edge/pkg/edged/edged.go]: dontainer -> container

Comment (arcanique) [reply edge/pkg/edged/edged.go]: ok

File: edge/pkg/edged/edged.go (hunk 11 above)
Comment (arcanique) [reply edge/pkg/edged/edged.go]: ok

**Pull request merged**